### DISCUSSION TITLE
Support offline mobile

### DISCUSSION THREAD

**Discussion by boxabirds** [coqui-ai/TTS#407] [General]
This repository is really great! Decent samples too. 

I see a huge opportunity for this to be extended to support mobile. 

There are a number of obstacles to this of course, including running on TF-Lite. 

If you ported it to Dart you could transpile it to iOS and Android.

**Comment by erogol** [comment]
Thx !  We also plan to expand :frog: TTS's use in offline and low-compute devices. 

Currently, some models we release are already real-time using simply the torch back-end. 

Targeting the mobile, we can make use of Torch Script, ONNX or TF-Lite export. So one step at a time we plan to get there. 

Actually you can already run the vanilla Tacotron on TF-Lite.  A bit of documentation is here https://github.com/coqui-ai/TTS/wiki/Converting-Torch-Tacotron-to-TF-2.0

**Reply by whitmer** [reply]
Is this the correct link? https://github.com/mozilla/TTS/wiki/Converting-Torch-to-TF-2.0

**Comment by Dmole** [comment]
There are currently 3 FOSS TTSEs for Mobile https://search.f-droid.org/?q=tts+engine&lang=en
A coqui-ai.apk could improve that short list.

**Comment by alfureu** [comment]
Any update on this? Just discovered this project and I am amazed this does not exist yet on Android...

**Reply by erogol** [reply]
No plans so far to support offline mobile.

**Reply by alfureu** [reply]
😢

**Comment by hb0nd** [comment]
I'm curious, why do so many TTS projects not bundle for Android? It seems strange, especially when TTS is most used on mobile devices

**Comment by Dmole** [comment]
Note that currently Coqui is less reliable, and orders of magnitude slower than Android TTS options.
Coqui would require a lot of work to use a reasonable amount of CPU/battery, and that's not currently the project focus.

**Comment by whitmer** [comment]
Just a bump for mobile support. I run an open source project that provides speech output tools for people with disabilities, and in the last year there have been multiple Android devices that are losing their native TTS options. We have many international users, and right now I have to pay a healthy chunk for a proprietary offline TTS provider in order to ensure support for everyone. I'd love to start routing some of that investment to something more open that can benefit more people. Open source offline as flexible and modern as Coqui would have a huge impact on the disability community.

**Comment by csukuangfj** [comment]
Please see
https://github.com/coqui-ai/TTS/discussions/3194

You can use [sherpa-onnx](https://github.com/k2-fsa/sherpa-onnx) to run VITS models from Coqui on Android and also embedded devices, e.g., raspberry pi.


We have pre-built Android APKs for the VITS English models from Coqui.
https://k2-fsa.github.io/sherpa/onnx/tts/apk.html

![image](https://github.com/coqui-ai/TTS/assets/5284924/56b2ec37-27de-40c0-8dd4-8245ae2ac8be)

**Reply by tentator** [reply]
WOWOWOWOW! Thanks a lot, those voices literally R O C K ! ! ! 
Bye bye IVONA and the like..

**Comment by a-pav** [comment]
I'm trying to understand how this project works, specially whether it's entirely offline or not. Well, which one is it? Do my text go through a online server before they're speaking?

**Reply by csukuangfj** [reply]
Please have a look at https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html

It is completely offline.

No data is sent out. Everything is processed locally.

No need to connect to a remote server.

Also, you don't need to be concerned about privacy issues.

**Reply by reviewzxy1** [reply]
> Please have a look at https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html
> 
> It is completely offline.
> 
> No data is sent out. Everything is processed locally.
> 
> No need to connect to a remote server.
> 
> Also, you don't need to be concerned about privacy issues.





Hi I have lot of questions, can I ask

1. Can this tts apps be used to read epub, pdf files ? 
      I wish to read light novels & got bored with googel & azure voices so looking for more

2. If yes how ? 
       , please give easy explanation I am dumb to use github & all that coding looking stuff

3. Are there any recommendation where can I get tts engines + voices, preffered offline
 Which can be integrated into system-accciblity

4.  Issue - There are many apps, but all have same name can't install 2 apps at once, help

**Reply by csukuangfj** [reply]
> Can this tts apps be used to read epub, pdf files ?
I wish to read light novels & got bored with googel & azure voices so looking for more

Yes, you can, as long as you have a corresponding reader app. Your app will invoke a text-to-speech engine.

What we provide is a text-to-speech engine that can be invoked by other APPs.


----

> If yes how ?
, please give easy explanation I am dumb to use github & all that coding looking stuff

Please search with google to download a reader app. After that, you can download our text-to-speech engine.
Then you can start your reader app and select our tts engine for text to speech.

---

> Are there any recommendation where can I get tts engines + voices, preffered offline
Which can be integrated into system-accciblity

Please download one from
https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html

Note that, there are quite a lot of APKs in the above page.

You can visit
http://huggingface.co/spaces/k2-fsa/text-to-speech
to listen to different voices and select one voice you like the most.

---

> Issue - There are many apps, but all have same name can't install 2 apps at once, help

Currently, you need to uninstall one to install a new one.

**Reply by reviewzxy1** [reply]
> > Can this tts apps be used to read epub, pdf files ?
> > I wish to read light novels & got bored with googel & azure voices so looking for more
> 
> Yes, you can, as long as you have a corresponding reader app. Your app will invoke a text-to-speech engine.
> 
> What we provide is a text-to-speech engine that can be invoked by other APPs.

®> (yeah just found out (tts server.apk) 

chinese app on github & it helped me select tts engines easily,) 
> > If yes how ?
> > , please give easy explanation I am dumb to use github & all that coding looking stuff
> 
> Please search with google to download a reader app. After that, you can download our text-to-speech engine. Then you can start your reader app and select our tts engine for text to speech.

®> > **I use "@ voice a loud reader " app for 3 years, found way to get voices workin, **

> > Are there any recommendation where can I get tts engines + voices, preffered offline
> > Which can be integrated into system-accciblity
> 
> Please download one from https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html
> 
> Note that, there are quite a lot of APKs in the above page.

®> question - (how can I find out which one is English) ¿¿

> You can visit http://huggingface.co/spaces/k2-fsa/text-to-speech to listen to different voices and select one voice you like the most.
> 
> > Issue - There are many apps, but all have same name can't install 2 apps at once, help
> 
> Currently, you need to uninstall one to install a new one.
😭😭
> I'll try do experiment of duplicate app

> Speaker Id. Had me confused for bit but later found how to use it's bit user unfriendly 

> Thanks for quick response bro 🙇

**Reply by reviewzxy1** [reply]
> > > Can this tts apps be used to read epub, pdf files ?
> > > I wish to read light novels & got bored with googel & azure voices so looking for more
> > 
> > 
> > Yes, you can, as long as you have a corresponding reader app. Your app will invoke a text-to-speech engine.
> > What we provide is a text-to-speech engine that can be invoked by other APPs.

 (yeah just found out (tts server.apk)
chinese app on github & it helped me select tts engines easily,)
> 
> > > If yes how ?
> > > , please give easy explanation I am dumb to use github & all that coding looking stuff
> > 
> > 
> > Please search with google to download a reader app. After that, you can download our text-to-speech engine. Then you can start your reader app and select our tts engine for text to speech.
> 
 **I use "@ voice a loud reader " app for 3 years, found way to get voices workin, **

> > > Are there any recommendation where can I get tts engines + voices, preffered offline
> > > Which can be integrated into system-accciblity
> > 
> > 
> > Please download one from https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html
> > Note that, there are quite a lot of APKs in the above page.
> 
 question - (how can I find out which one is English) ¿¿ 
Found out , Voices which has "US" in it 
 (how do you know which app has more speaker Id, 
>Is there way to know without going to Web, or I >need to remember it
> > You can visit http://huggingface.co/spaces/k2-fsa/text-to-speech to listen to different voices and select one voice you like the most.
> > > Issue - There are many apps, but all have same name can't install 2 apps at once, help
> > 
> > 
> > Currently, you need to uninstall one to install a new one.
 😭😭
 I'll try do experiment of duplicate app
> 
Speaker Id. Had me confused for bit but later found how to use it's bit a lot user unfriendly
> 
Thanks for quick response bro 🙇

**Reply by csukuangfj** [reply]
> question - (how can I find out which one is English)

Please read the description in the page
https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html
carefully.

`en` is short for English. Any APK supporting English will have `en` in the filename. It should be straightforward to tell
which APK supports English.

**Reply by reviewzxy1** [reply]
Do you know how to find out how many speaker Id, app has? Without going to web ? 

- sory for asking 😅 I found out how to do it, app has written how many speaker Id it has



Thanks en short name guidance

Curious are developer / manager of this tts
